### DISCUSSION TITLE
使用中的两个问题，一、识别错了的字幕如何人工修正后放到剪辑字幕里？二、经常报错无法剪辑，下面有举例

### DISCUSSION THREAD

**Discussion by joe20002000** [modelscope/FunClip#128] [Q&A] [answered]
**1. 例如，我asr了一个14分钟的讲课视频，很多语音识别错误，能否人工修正后，然后根据修正的文本进行剪辑，然后剪辑后的字幕为修正 的。
2. 对视频不剪辑，所有文本复制进去无法生成带字幕的视频，多处分段剪辑也经常无法生成带字幕的视频，最后反复尝试，才加入两处#符合，才能生成完整视频带字幕。目前无法确定是什么原因，没有规律性，想找见规律生成剪辑或不剪辑带字幕的视频，无法保障，本来是想提高效率，如果这样就会令人望而却步了。
下面是asr后的识别文本（里面加了两个#才能生成完整字幕的视频）：**
这次课我们来学习第一章，第三节有关分子结构与结构式的相关内容。在上一次课，我们了解了有机化学就是一门碳化合物的化学，并且了解了有机化合物与人类关系密切，有机化合物的性质特点呢不同于无机物，主要表现在五个方面，依然熔沸点、低难溶于水，反应速率慢，副反应产物多等。第三个方面呢，有机化合物的数量众多啊，它的数量众多多到哪种程度呢？有机化合物，有上千万种啊，有具体数字表明他有八千多万种，八十多。而我们的无机化后呢，有几百万种啊，为什么不确定是几百呢？是一百还是两百？#我们在上次课提到上一个PPT的讲到，它是不到四十四十万的啊，在这几百万中呢，它包含了配位化合物。#在无机化学课程中，我们已经要学过了。那这个配位化合中心离子是金属金属离子作为配配列中心，就那些，而胚体绝大多数都是有机化合物啊，那么无机化合物呢？如果把这些配配化合物物因有机结构不同，引引而得到的配位化合物算起来也就几百万种来有机化学家把这一类物质呢称为金属有机化合物，在无机物中把它称为配位化合物，每年新增的有机化后呢也有几千种，为什么有机物的数量比无机物多这么多呢？大家想一想，什么原因，有机物的组成元素还比较少。嗯，碳氢氧氮、硫、磷卤素啊就十几种，而无机物的组成元素种类很多，有一百种左右啊，一百种左右。嗯，有机物的数量为什么反而比无机物多这么多呢？主要原因就是由于有机化合物，结构繁杂嗯，组成元素虽然比较简单，但它的结构繁杂，这也是为什么一个碳元素就能成为成为一门独立的学科的原因。我们这一次课呢主要了解有机分子的结构与结构式的表示分子。那么在我们的化学课中啊，基础化学课包括中学的以及大大学上学期，我们学到的无机化学都知道分子是由原子当然也包括离子啊，按照一定排列顺序、相互结合、相互影响及格而形成的长体这种排列顺序和相互作用关系就称为分子结构。然后在这儿给了一个相对简单的有机分子乙烷啊，我们能看到碳碳这两个元素以及这两个原子以及这六个氢原子按照一定顺序连接而成。那么分子的性质呢不仅取决于组成分子的原子种类原子的数量，更重要的取决于分子的结构啊，有机化合物种类繁多。其实主要原因就是由于它的分子结构比较复杂。那有些同学应该也知道中学学过的啊同分异构现象就是造成分子结构负杂一种表现。那么接下来我们来看一下分子的异构现象，分子式都是碳二氢氯一氧。这样的画横就有两种，一种是乙醇，另外一个是假迷啊。另外一个是钱亚宾。这种由于分子式相同，结构不同而造成性质不同的现象就称为同分异构，是由分子之间间隔次序不同而产生的异构现象。我们把它也称为构造异构啊，也称为构造移动。我们一会儿会讲到同分异构的分类啊，同非异构体的分类，其中一种就是我们的第二性的构造异构。啊，其实除了构造异构之外呢，还有多种异构方式啊，比如说关能团异构呀等等。随着组成分子的原子数目的增加，构造异构体的数目不断的增加，激烈的在增加，那么一会儿会有具体的数据数据来给大家展示。啊，比如在这儿给了一个碳似氢石氧，就包含七种异构体。那同分异构这样一种现象有哪些分类呢？我们来看到同分异构体的，从大的方面主要分成三种。第一种是构造异构体，构造易构体呢有五个方面啊，有碳加异构体，which异构体这位置是谁呀？可以是关能团的，也可以是取代机位置不同而造成的异构啊，比如说主链是五个碳的玩厅，在侧链上呢是个甲基，这个甲级可以连在二号位，也可以连在三号位。哎，这个就是位置移动，还有我们的丙醇，三个碳依次相连。在碳上连有一个羟基，这就是丙醇的结构。用这个抢击连在一号台二号碳就形成了两种不同结构的丙醇，这就是位置移动啊。第三种呢是官能团异构体。那这关能团一勾起，我相信大家能说出很多来啊，比如说丙醛和丙酮，当然还包括我们刚才已经说到的乙醇和甲醚相同碳原子数目的一元梭酸，以及对应碳原子数目的酯，这就构成了官能统一构嗯关注成一个。除此之外呢，还有互变异构体、嫁界异构体等等。啊，在后面的症节中，以及我们在高等有机化学中还会学到第二大类呢，就是立体异构体的立体异构体。那么在后面呢，我们还会具体的讲到，它主要包含了包含了构想异构体构象异构体啊，这个构象异构体呢哎我们再下一章很快就会学到构型异构异构呢在我们的立体化学印章中会学习到啊，当然在西天议章中也会涉及到构型异构体中的一种叫几何异构体。在立体化学异程中，主要学习的是玄光异构体。第三大类是电子互变异构体啊，在咱们基础有机化学课程中呢，有关电子互变异构体呢呃，讲的相对比较少，相对比较少。但我们要了解一下。好，后面再来看给出的同飞异度现象数量的变化啊，碳原子数是四个的时候的对应的烷厅。以烷听为例，它的异构体的数是两个，五个碳的时候是三个通分。一个问题，碳原子数目是六对应顽厅的同分异构体现象。五个那碳二十氢四十二这样一种蓝厅对应的同分异构体数目就达到了三十六万多重，看看是不是组成分子的原子数目在增加，同非异构体的数量在激增嗯，激增不是慢慢的增加啊，而是可以说几何几数的这样一种增转方式。那么这么多的有机化合物又如何来表达它们呢？显然，仅仅通过分子式是不能很好的准确的表达不同，有机分子它们对应的结构不同，那不能表达这样一种现象。那怎么来表达呢？我们接下来了解有机分子结构式的表示，主要涉及到五种方法啊，主要涉及到五种方法，那么它们的优劣大家一目了然。那我们现在先来看第一种路易斯十，它是用一对电子来表示一个共价键。一对电子来表示一个共价键。那么在中学实际上我们已经用这样一种路易四十表达梗多种分子。比如说氧气，比如说氢二氧没有，当然还包括我们的氯化氢啊、氯化氢之间共用一对电子。哎，我们用这个电子式电子路易四世来表示花点，或者差一个电子，就用一个点或者差来表示。当然这都指的是最外层。第二种方式呢是开口乐视，它是一个短线来表达表示一个共价键。很显然，用一只短线来表达共价键比用两个点或者一个点，一个叉或者两个叉来表示共价键就要简单多。我底下举了例子，大家一看就知道哪个书写更方便。对于甲烷来说，用路易四世来表示就是这样一种情况，开普勒式的这样一种写法。还有乙炔是不是开普勒世要比路易斯是表示这样对应有机分子的结构要简单多了。那么第三种方式呢是建现式是界限，呃，是蛛网肾啊，是蛛网肾，最简单的是界限式啊。除此之外呢，还有第四种的缩写式。大家一看对比就知道它们的不同。很显然，界限式表示有机分子的结构简单明了，书写方便。再回想一下，我们在上一节上一讲讲到的那些药物分子，还记得吧？还有点印象吧啊，我心里氯霉素、青霉素如果用刚才的开普勒时，路易斯是蛛网式这三种方式来写。你写一下那个复杂的结构，要写到什么时候去这件事？就是我们在今后学习有机化学书写一个有机化合物体构造时，主要的表示方式，主要的表示方式。那么在这儿需要强调的是，这姐姐给出的是构造式，它它表达有些分子呢也有一些缺陷啊，比如说不能表达立体结构啊，我们在后面立体化学一章中会讲到啊，还有其他的方式，比如说楔形式、透视式、嗯，非渲染投稿师等等。咱具体的章节，我们再来学习。好，这次课到这儿。

界面提示：
![image](https://github.com/user-attachments/assets/c98c61c9-85f8-41df-a51d-b7130f81b530)

**Comment by LauraGPT** [comment, marked as answer]
这两个问题我在当前 `main`（`80ea6ea`）上重新复现了，结论如下。

### 1. 目前不能直接用“人工修正后的文字”定位剪辑并自动烧入修正字幕

“待裁剪文本”是**片段定位框**，不是字幕编辑框。它会在原始 ASR token 中查找输入文字来取得时间戳；“裁剪+字幕”使用的也仍是原始识别句子。因此：

- 把错字改正确后再放入“待裁剪文本”，新文字与 ASR 原文不同，通常会匹配不到。
- 即使成功剪辑，当前界面烧入的仍是 ASR 原字幕，不会自动采用修改后的文字。

当前可靠流程是：

1. 使用“识别结果”中的**原文**选择片段；每段选一段连续、唯一、长度适中的原文。
2. 先点“裁剪”，取得视频和“裁剪部分 SRT 字幕内容”。
3. 将 SRT 保存为 `corrected.srt`，只修改字幕文字，保留编号和时间戳。
4. 将修正后的 SRT 烧入已经剪好的视频：

```bash
ffmpeg -i clipped.mp4 \
  -vf "subtitles=corrected.srt:fontsdir=/usr/share/fonts/opentype/noto" \
  -c:v libx264 -c:a copy clipped_corrected.mp4
```

`fontsdir` 请按本机字体目录调整。若只是想给完整的 14 分钟视频修正字幕，不需要把全文粘到“待裁剪文本”：直接下载完整 SRT、修改文字后，用同一命令烧入原视频即可。

### 2. `#` 的规则是“多段片段分隔符”

例如：

```text
第一段原识别文字#第二段原识别文字
```

FunClip 会分别查找两段文字对应的时间区间，再按输入顺序拼接。你截图中的日志：

```text
from 0.98 to 68.565, from 80.49 to 844.0
```

表示实际找到了两个片段并进行了拼接；`68.565` 到 `80.49` 秒之间的内容没有包含，所以 `#` 并不是“生成完整视频”的开关。

为减少“无法剪辑”：

- 不要把 14 分钟全文作为一个查询；只复制真正需要的、连续的原识别文字。
- 多个不连续片段用 `#` 分开。
- 不要先改字、增加标题或换行后再匹配。
- 避免使用会在全文中重复出现的短句，尽量选更长且唯一的原文。

另外，当前代码对“同一文字命中多次且没有偏移参数”的情况确实存在一个可复现异常，并不全是操作问题；修复前使用更长、唯一的原文可以绕开。

我用 4 秒测试视频做了完整验证：两个由 `#` 分隔的片段成功生成 2 秒拼接视频，SRT 时间轴重排正确；修改 SRT 后再烧录，输出仍为 2 秒且字幕只出现在画面底部。

**Comment by LauraGPT** [comment]
补充进展：上面提到的“同一文字命中多次、且没有偏移参数时崩溃”已经提交修复 PR：

https://github.com/modelscope/FunClip/pull/177

根因是音频和视频两条裁剪路径都只在带 `[...]` 偏移时定义 `match`，普通重复查询返回多个时间段后却仍读取该变量，触发 `UnboundLocalError`。修复保持原有语义：重复文字仍会匹配并拼接全部命中，只是不再崩溃，也不会让前一个 `#` 片段的偏移匹配状态泄漏到下一个片段。

验证包括：

- 修复前音频/视频两条回归都复现同一个异常，修复后 2/2 通过；
- 仓库测试 16 项通过、1 项可选在线测试跳过；
- 4 秒 H.264/AAC 合成视频中，同一个“重复”查询正确剪出 `0.5-1.0s` 和 `2.5-3.0s`，输出为 1.000 秒，字幕时间轴重排为两个连续的 0.5 秒片段。

PR 合并前仍建议优先选择较长且唯一的原识别文字；合并后重复短句至少会稳定返回所有命中，不再因这个局部变量异常失败。

**Comment by LauraGPT** [comment]
修复已合并到 FunClip `main`：

- PR: https://github.com/modelscope/FunClip/pull/177
- merge commit: [`d30dacc`](https://github.com/modelscope/FunClip/commit/d30daccff432534788bc7a375091d33513465ab8)

现在普通短句即使在识别结果中重复出现，也会稳定剪出并拼接全部命中，不再触发 `UnboundLocalError`。带 `[开始偏移, 结束偏移]` 的重复查询也会在最终裁剪信息中显示已格式化的命中数量提示；直接传入时间戳的裁剪路径不再依赖文本匹配状态。

更新到最新 `main` 后即可使用。字幕人工修正的边界仍与已接受答案一致：先按原识别文字剪出片段，再修改剪出后的 SRT；当前界面不会自动把修正文本替换为定位文本和烧录字幕。
